components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/ciba/OAuth2CibaEndpoint.java

```diff
@@ -0,0 +1,122 @@
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+@Path("/ciba")
+public class OAuth2CibaEndpoint {
+
+    private static final Log log = LogFactory.getLog(OAuth2CibaEndpoint.class);
+
+    @POST
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("application/json")
+    public Response ciba(@Context HttpServletRequest request, @Context HttpServletResponse response)
+            throws  CibaClientException {
+
+        // Capture all  Authentication Request parameters.
+        Map<String, String[]> requestParameterMap = request.getParameterMap();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Authentication request has hit Client Initiated Back-channel Authentication EndPoint.");
+        }
+
+        try {
+            if (!requestParameterMap.containsKey(CibaConstants.REQUEST)) {
+                // Mandatory 'request' parameter does not exist.
+
+                if (log.isDebugEnabled()) {
+                    log.debug("CIBA Authentication Request that hits Client Initiated Authentication Endpoint has " +
+                            "no 'request' parameter.");
+                }
+                throw new CibaAuthFailureException(OAuth2ErrorCodes.INVALID_REQUEST,
+                        "missing the mandated parameter : (request)");
+            }
+
+            // Capturing authentication request.
+            String authRequest = request.getParameter(CibaConstants.REQUEST);
+
+            if (log.isDebugEnabled()) {
+                log.debug("CIBA Authentication Request with  'request' :" + authRequest + "  has hit Client " +
+                        "Initiated Back-Channel Authentication EndPoint.");
+            }
+
+            CibaAuthRequestValidator.getInstance().validateRequest(authRequest);
+
+            CibaAuthRequestValidator.getInstance().validateClient(authRequest);
+            // The CIBA Authentication Request is with proper client.
+
+            CibaAuthRequestValidator.getInstance().validateUserHint(authRequest);
+            // The CIBA Authentication Request is with proper user hint.
+
+            // Validate Authentication request.
+            CibaAuthRequestValidator.getInstance().validateAuthRequestParams(authRequest);
+
+            // Prepare RequestDTO with validated parameters.
+            CibaAuthCodeRequest cibaAuthCodeRequest =
+                    CibaAuthRequestValidator.getInstance().prepareRequestDTO(authRequest);
+
+            // Obtain Response from service layer of CIBA.
+            CibaAuthCodeResponse cibaAuthCodeResponse = null;
+            try {
+                cibaAuthCodeResponse = EndpointUtil.getCibaAuthService().generateAuthCodeResponse(cibaAuthCodeRequest);
+            } catch (CibaCoreException e) {
+                throw new CibaAuthFailureException(OAuth2ErrorCodes.SERVER_ERROR, "Error while generating " +
```

Comment:
Since this is the endpoint, we will have to log the **server errors** with their stack trace.

Reply:
cant we catch all errors and log it there [in preparing error response catch error and log if it is a server error]